packages/@aws-cdk/aws-elasticloadbalancingv2/lib/alb/application-listener.ts

```diff
@@ -143,8 +147,20 @@ export class ApplicationListener extends BaseListener implements IApplicationLis
   /**
    * Add one or more certificates to this listener.
    */
-  public addCertificateArns(_id: string, arns: string[]): void {
-    this.certificateArns.push(...arns);
+  public addCertificateArns(id: string, arns: string[]): void {
+    const [first, ...rest] = arns;
+    const additionalCertArns = this.certificateArns.length ? arns : rest;
+
+    if (!this.certificateArns.length && typeof first === "string") {
+      this.certificateArns.push(first);
+    }
+
+    if (additionalCertArns.length) {
+      new ApplicationListenerCertificate(this, id, {
+        listener: this,
+        certificateArns: additionalCertArns
+      });
+    }
```

Comment:
I agree that the whole logic here is a little hairy to handle different cases. What if we embraced mutability, and do something like this:

```ts
addCertificateArns(id, arns: string[]) {
  if (arns.length > 0 && this.certificateArns.length === 0) {
    // Pop off the first certificate and put it on the main resource
    const first = arns.splice(0, 1)[0];
    this.certificateArns.push(first);
  } 

  if (arns.length > 0) { 
    new ApplicationListenerCertificate(this, id, { ... });
  }
}
```

If you feel bad about mutating the input argument (reasonable! :) then we can always copy it into a local variable.